src/crypto/DeviceList.js

```diff
@@ -0,0 +1,446 @@
+    }
+
+    /**
+     * Download the keys for a list of users and stores the keys in the session
+     * store.
+     * @param {Array} userIds The users to fetch.
+     * @param {bool} forceDownload Always download the keys even if cached.
+     *
+     * @return {Promise} A promise which resolves to a map userId->deviceId->{@link
+     * module:crypto/deviceinfo|DeviceInfo}.
+     */
+    downloadKeys(userIds, forceDownload) {
+        const self = this;
+
+        // promises we need to wait for while the download happens
+        const promises = [];
+
+        // list of userids we need to download keys for
+        let downloadUsers = [];
+
+        function perUserCatch(u) {
+            return function(e) {
+                console.warn('Error downloading keys for user ' + u + ':', e);
+            };
+        }
+
+        if (forceDownload) {
+            downloadUsers = userIds;
+        } else {
+            for (let i = 0; i < userIds.length; ++i) {
+                const u = userIds[i];
+
+                const inprogress = this._keyDownloadsInProgressByUser[u];
+                if (inprogress) {
+                    // wait for the download to complete
+                    promises.push(q.any(inprogress).catch(perUserCatch(u)));
+                } else if (!this.getStoredDevicesForUser(u)) {
+                    downloadUsers.push(u);
+                }
+            }
+        }
+
+        if (downloadUsers.length > 0) {
+            const r = this._doKeyDownloadForUsers(downloadUsers);
+            downloadUsers.map(function(u) {
+                promises.push(r[u].catch(perUserCatch(u)));
+            });
+        }
+
+        return q.all(promises).then(function() {
+            return self._getDevicesFromStore(userIds);
+        });
+    }
+
+    /**
+     * Get the stored device keys for a list of user ids
+     *
+     * @param {string[]} userIds the list of users to list keys for.
+     *
+     * @return {Object} userId->deviceId->{@link module:crypto/deviceinfo|DeviceInfo}.
+     */
+    _getDevicesFromStore(userIds) {
+        const stored = {};
+        const self = this;
+        userIds.map(function(u) {
+            stored[u] = {};
+            const devices = self.getStoredDevicesForUser(u) || [];
+            devices.map(function(dev) {
+                stored[u][dev.deviceId] = dev;
+            });
+        });
+        return stored;
+    }
+
+    /**
+     * Get the stored device keys for a user id
+     *
+     * @param {string} userId the user to list keys for.
+     *
+     * @return {module:crypto/deviceinfo[]|null} list of devices, or null if we haven't
+     * managed to get a list of devices for this user yet.
+     */
+    getStoredDevicesForUser(userId) {
+        const devs = this._sessionStore.getEndToEndDevicesForUser(userId);
+        if (!devs) {
+            return null;
+        }
+        const res = [];
+        for (const deviceId in devs) {
+            if (devs.hasOwnProperty(deviceId)) {
```

Comment:
bit of a shame that we're doing a weird hybrid of `const` and then horrid `hasOwnProperty` crap. but this can be fixed up in a later hygiene pass.....